### PR TITLE
Booster implementation

### PR DESCRIPTION
**Pull request by octotentakel**

### CODE REVIEW

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/BoostMeterUiFx.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * @author PTS34A
+ */
+public class BoostMeterUiFx extends Drawable<Car> {
+
+    /**
+     * Initiates a new NotificationUiFx Object using the given parameters.
+     *
+     * @param canvas The canvas on which this Notification is placed.
+     */
+    public BoostMeterUiFx(GameCanvas canvas, Car car) {
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/BoostMeterUiFx.java]: MINOR Document the parameter(s): car [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/ScoreBoardUiFx.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+     *
+     * @param canvas The canvas on which this Notification is placed.
+     */
+    public ScoreBoardUiFx(GameCanvas canvas, Game game) {
+        super(canvas, game);
+    }
+
+    @Override
+    public void draw(GraphicsContext context) {
+        float width, height, x, y;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/ScoreBoardUiFx.java]: MINOR Declare "height" on a separate line. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1659)
MINOR Declare "x" on a separate line. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1659)
MINOR Declare "y" on a separate line. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1659)

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/ScoreBoardUiFx.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        Game game = super.getModel();
+
+        // Prepare text to display
+        String text = String.format("%d - %s - %d", game.getEventCountByType(EventType.GOAL_BLUE), game.getTimeLeftString(), game.getEventCountByType(EventType.GOAL_RED));
+
+        width = PhysicsUtilities.toPixelWidth(DisplayConstants.SCOREBOARD_WIDTH);
+        height = PhysicsUtilities.toPixelHeight(DisplayConstants.SCOREBOARD_HEIGHT);
+        x = PhysicsUtilities.toPixelX((float) (game.getMap().getSize().getWidth() / 2) - (DisplayConstants.SCOREBOARD_WIDTH / 2));//PhysicsUtilities.toPixelX(game.getX());
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/ScoreBoardUiFx.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/ScoreBoardUiFx.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * @author PTS34A
+ */
+public class ScoreBoardUiFx extends Drawable<Game> {
+
+    /**
+     * Initiates a new NotificationUiFx Object using the given parameters.
+     *
+     * @param canvas The canvas on which this Notification is placed.
+     */
+    public ScoreBoardUiFx(GameCanvas canvas, Game game) {
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/ScoreBoardUiFx.java]: MINOR Document the parameter(s): game [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java [hunk 5]
```diff
@@ -102,4 +109,37 @@ private void drawWheel(WheelPhysics wheel, GraphicsContext gc) {
         gc.restore();
     }
 
+    private void drawBoostTrail(List<Point2D> trail, GraphicsContext gc) {
```
Comment (ghost) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java:103]: MAJOR Remove this unused private "drawBoostTrail" method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java [hunk 6]
```diff
@@ -102,4 +109,37 @@ private void drawWheel(WheelPhysics wheel, GraphicsContext gc) {
         gc.restore();
     }
 
+    private void drawBoostTrail(List<Point2D> trail, GraphicsContext gc) {
+
+        double prevX, prevY, x, y;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java]: MINOR Declare "prevY" on a separate line. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1659)
MINOR Declare "x" on a separate line. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1659)
MINOR Declare "y" on a separate line. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1659)

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java [hunk 7]
```diff
@@ -102,4 +109,37 @@ private void drawWheel(WheelPhysics wheel, GraphicsContext gc) {
...
+    private void drawBoostTrail(List<Point2D> trail, GraphicsContext gc) {
+
+        double prevX, prevY, x, y;
+        double o = 0.0;
+        int r = 30;
+        int g = 30;
+        int b = 230;
+
+        // Cancel when there is no trail.
+        if (trail.size() == 0) {
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java]: MAJOR Use isEmpty() to check whether the collection is empty or not. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1155)

File: Soccar [UI]/src/nl/soccar/ui/fx/controller/GameViewFXMLController.java [hunk 8]
```diff
@@ -6,24 +6,18 @@
 import javafx.fxml.Initializable;
 import javafx.scene.canvas.Canvas;
 import javafx.scene.layout.AnchorPane;
-import nl.soccar.library.Car;
-import nl.soccar.library.Map;
-import nl.soccar.library.Session;
+import nl.soccar.library.*;
+import nl.soccar.library.enumeration.Duration;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/controller/GameViewFXMLController.java]: MINOR Remove this unused import 'nl.soccar.library.enumeration.Duration'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java [hunk 9]
```diff
@@ -5,13 +5,19 @@
...
 import javafx.util.Duration;
+import nl.soccar.library.Event;
 import nl.soccar.library.Game;
+import nl.soccar.library.Notification;
+import nl.soccar.library.enumeration.GameStatus;
+import nl.soccar.physics.PhysicsConstants;
+import nl.soccar.ui.DisplayConstants;
 import nl.soccar.ui.drawable.Drawable;
 import nl.soccar.ui.drawable.GameCanvas;
+import nl.soccar.ui.fx.drawable.NotificationUiFx;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java]: MINOR Remove this unused import 'nl.soccar.ui.fx.drawable.NotificationUiFx'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java [hunk 10]
```diff
@@ -5,13 +5,19 @@
...
 import nl.soccar.ui.input.Keyboard;
 
+import java.time.LocalTime;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java]: MINOR Remove this unused import 'java.time.LocalTime'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java [hunk 11]
```diff
@@ -5,13 +5,19 @@
 import javafx.scene.canvas.Canvas;
 import javafx.scene.canvas.GraphicsContext;
 import javafx.util.Duration;
+import nl.soccar.library.Event;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java]: MINOR Remove this unused import 'nl.soccar.library.Event'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java [hunk 12]
```diff
@@ -5,13 +5,19 @@
 import javafx.scene.canvas.Canvas;
 import javafx.scene.canvas.GraphicsContext;
 import javafx.util.Duration;
+import nl.soccar.library.Event;
 import nl.soccar.library.Game;
+import nl.soccar.library.Notification;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java]: MINOR Remove this unused import 'nl.soccar.library.Notification'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java [hunk 13]
```diff
@@ -5,13 +5,19 @@
 import javafx.scene.canvas.Canvas;
 import javafx.scene.canvas.GraphicsContext;
 import javafx.util.Duration;
+import nl.soccar.library.Event;
 import nl.soccar.library.Game;
+import nl.soccar.library.Notification;
+import nl.soccar.library.enumeration.GameStatus;
+import nl.soccar.physics.PhysicsConstants;
+import nl.soccar.ui.DisplayConstants;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/GameCanvasFx.java]: MINOR Remove this unused import 'nl.soccar.ui.DisplayConstants'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by TehMessiah** [comment]
SonarQube analysis reported 18 issues
* MAJOR 4 major
* MINOR 14 minor

Watch the comments in this conversation to review them.

**Comment by BeeHiveJava** [comment]
retest this please

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java (hunk 5 above)
Comment (octotentakel) [reply Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java:103]: This method is used in the draw method of the same class.

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java [hunk 14]
```diff
@@ -102,4 +109,42 @@ private void drawWheel(WheelPhysics wheel, GraphicsContext gc) {
...
 
+    private void drawBoostTrail(List<Point2D> trail, GraphicsContext gc) {
+
+        double prevX;
+        double prevY;
+        double x;
+        double y;
+        double o = 0.0;
+
+        // TODO: Boost color should be the color of the team the player is in.
```
Comment (ghost) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java:111]: INFO Complete the task associated to this TODO comment. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1135)

**Comment by TehMessiah** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

**Review by Ranray — APPROVED** [review]

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java [hunk 15]
```diff
@@ -72,9 +78,9 @@ public void draw(GraphicsContext context) {
             // TODO : Change location of sending this message (Ugly implementation for now)
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java]: INFO Complete the task associated to this TODO comment. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1135)

**Comment by ghost** [comment]
SonarQube analysis reported 3 issues
* MAJOR 1 major
* INFO 2 info

Watch the comments in this conversation to review them.

File: Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java [hunk 16]
```diff
@@ -118,4 +100,64 @@ private void drawWheel(WheelPhysics wheel, GraphicsContext gc) {
...
+        if (ClientController.getInstance().getCurrentPlayer().getUsername().equals(super.getModel().getPlayer().getUsername())) {
+            SteerAction steerAction = Keyboard.getSteerAction();
+            HandbrakeAction handbrakeAction = Keyboard.getHandbrakeAction();
+            ThrottleAction throttleAction = Keyboard.getThrottleAction();
+
+            super.getModel().setSteerAction(steerAction);
+            super.getModel().setHandbrakeAction(handbrakeAction);
+            super.getModel().setThrottleAction(throttleAction);
+
+            // TODO : Change location of sending this message (Ugly implementation for now)
```
Comment (ghost) [inline Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java:154]: INFO Complete the task associated to this TODO comment. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1135)

**Pull request merged**